yarn-project/simulator/src/private/providers/circuit_recording/simulation_provider_recorder_wrapper.ts

```diff
@@ -0,0 +1,85 @@
+import type { NoirCompiledCircuitWithName } from '@aztec/stdlib/noir';
+
+import type { ACIRCallback, ACIRExecutionResult } from '../../acvm/acvm.js';
+import type { ACVMWitness } from '../../acvm/acvm_types.js';
+import type { SimulationProvider } from '../simulation_provider.js';
+import { CircuitRecorder } from './circuit_recorder.js';
+
+/**
+ * Takes a simulation provider and wraps it in a circuit recorder. See CircuitRecorder for more details on how circuit
+ * recording works.
+ */
+export class SimulationProviderRecorderWrapper implements SimulationProvider {
+  constructor(private simulator: SimulationProvider, private log = createLogger('simulator-recorder')) {}
+
+  async executeProtocolCircuit(
+    input: ACVMWitness,
+    artifact: NoirCompiledCircuitWithName,
+    callback: ForeignCallHandler | undefined,
+  ): Promise<ACVMWitness> {
+    const recordingEnabled = process.env.CIRCUIT_RECORD_DIR !== undefined;
+    let circuitRecorder: CircuitRecorder | undefined;
+    let wrappedCallback: ForeignCallHandler | undefined = undefined;
+    if (recordingEnabled) {
+      // Recording is enabled so we start a new circuit recorder
+      circuitRecorder = await CircuitRecorder.start(
+        process.env.CIRCUIT_RECORD_DIR!,
+        input,
+        Buffer.from(artifact.bytecode, 'base64'),
+        artifact.name,
+        'main',
+      );
+      if (callback !== undefined) {
+        // Callback was provided so we wrap it in a circuit recorder callback wrapper
+        wrappedCallback = circuitRecorder.wrapProtocolCircuitCallback(callback);
+      }
+    }
+
+    try {
+      const result = await this.simulator.executeProtocolCircuit(
+        input,
+        artifact,
+        wrappedCallback ? wrappedCallback : callback,
+      );
+
+      if (recordingEnabled) {
+        // Recording is enabled and witness generation is complete so we finish the circuit recorder
+        await circuitRecorder!.finish();
+      }
+
+      return result;
+    } catch (err) {
+      this.log.error('Circuit execution failed', { error: err });
+      throw err;
+    }
+  }
+
+  async executeUserCircuit(
+    input: ACVMWitness,
+    artifact: FunctionArtifactWithContractName,
+    callback: ACIRCallback,
+  ): Promise<ACIRExecutionResult> {
+    if (!process.env.CIRCUIT_RECORD_DIR) {
+      return this.simulator.executeUserCircuit(input, artifact, callback);
+    }
+
+    // The CIRCUIT_RECORD_DIR environment variable is set, so we start a new circuit recorder
+    const recorder = await CircuitRecorder.start(
+      process.env.CIRCUIT_RECORD_DIR,
+      input,
+      artifact.bytecode,
+      artifact.contractName,
+      artifact.name,
+    );
+    // We wrap the oracle callback in a circuit recorder callback wrapper
+    const wrappedCallback = recorder.wrapUserCircuitCallback(callback);
+    const result = await this.simulator.executeUserCircuit(input, artifact, wrappedCallback);
```

Comment:
What if this throws?

Reply:
Addressed in https://github.com/AztecProtocol/aztec-packages/pull/12148/commits/732554c700688a9ec3f3335c5572fdc2d4324152

Now if it throws an error gets added to the file and then the error gets re-thrown
![image](https://github.com/user-attachments/assets/675870a0-b512-4236-ba60-e1c0b0456a16)